hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java

```diff
@@ -82,39 +84,95 @@ public WanReplicationServiceImpl(Node node) {
+        }
+
+        ConcurrentMap<String, WanReplicationPublisher> publishers = createConcurrentHashMap(publisherCount);
+        Map<String, AbstractWanPublisherConfig> publisherConfigs = createHashMap(publisherCount);
+
+        customPublisherConfigs.forEach(
+                publisherConfig -> {
+                    String publisherId = getPublisherIdOrGroupName(publisherConfig);
+                    if (publishers.containsKey(publisherId)) {
+                        throw new InvalidConfigurationException(
+                                "Detected duplicate publisher ID '" + publisherId + "' for a single WAN replication config");
+                    }
+
+                    WanReplicationPublisher publisher = createPublisher(publisherConfig);
+                    publishers.put(publisherId, publisher);
+                    publisherConfigs.put(publisherId, publisherConfig);
+                });
+
+        for (Entry<String, WanReplicationPublisher> publisherEntry : publishers.entrySet()) {
+            String publisherId = publisherEntry.getKey();
+            WanReplicationPublisher publisher = publisherEntry.getValue();
+            node.getSerializationService().getManagedContext().initialize(publisher);
+            publisher.init(wanConfig, publisherConfigs.get(publisherId));
+        }
+
+        return publishers;
+    }
+
+    /**
+     * Instantiates a {@link WanReplicationPublisher} from the provided publisher
+     * configuration.
+     *
+     * @param publisherConfig the WAN publisher configuration
+     * @return the WAN replication publisher
+     * @throws InvalidConfigurationException if the method was unable to create the publisher because there was no
+     *                                       implementation or class name defined on the config
+     */
+    private WanReplicationPublisher createPublisher(AbstractWanPublisherConfig publisherConfig) {
+        WanReplicationPublisher publisher = getOrCreate(
+                (WanReplicationPublisher) publisherConfig.getImplementation(),
+                node.getConfigClassLoader(),
+                publisherConfig.getClassName());
+        if (publisher == null) {
+            throw new InvalidConfigurationException("Either \'implementation\' or \'className\' "
+                    + "attribute need to be set in the WAN publisher configuration for publisher " + publisherConfig);
         }
+        return publisher;
+    }
 
-        return targetEndpoints;
+    /**
+     * Returns the publisher ID for the given WAN publisher configuration which
+     * is then used for identifying the WAN publisher in a WAN replication
+     * scheme.
+     * If the publisher ID is empty, returns the publisher group name.
+     *
+     * @param publisherConfig the WAN replication publisher configuration
+     * @return the publisher ID or group name
+     */
+    public static @Nonnull
+    String getPublisherIdOrGroupName(AbstractWanPublisherConfig publisherConfig) {
```

Comment:
can we call this `getWanPublisherId`? since in the end we are concerned with an id. It is group name or something else.

Reply:
Right, will change.

Reply:
Addressed in a separate commit.